File: LeetCode_Monthly_Challenge/July/C++/Week_1/Binary Tree Level Order Traversal II.cpp
/*
PROBLEM:
Given a binary tree, return the bottom-up level order traversal of its nodes' values. (ie, from left to right, level by level from leaf to root).
For example:
Given binary tree [3,9,20,null,null,15,7],
    3
   / \
  9  20
    /  \
   15   7
return its bottom-up level order traversal as:
[
  [15,7],
  [9,20],
  [3]
]
*/


/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
 
 
class Solution {
public:
    vector<vector<int>> levelOrderBottom(TreeNode* root) {
        vector<vector<int>> result;
        if(!root)
            return result;
        queue<TreeNode*> q;
        q.push(root);
        while(!q.empty()){
            int n=q.size();
            vector<int> nodes(n);
            for(int i=0;i<n;i++){
                TreeNode* node = q.front();
                q.pop();
                nodes[i]=node->val;
                if(node->left) q.push(node->left);
                if(node->right) q.push(node->right);
                
            }
            result.push_back(nodes);
        }
        reverse(result.begin(),result.end());
        return result;
    }
};

File: LeetCode_Monthly_Challenge/July/Python3/Week_1/Binary Tree Level Order Traversal II.py
"""
PROBLEM:
Given a binary tree, return the bottom-up level order traversal of its nodes' values. (ie, from left to right, level by level from leaf to root).
For example:
Given binary tree [3,9,20,null,null,15,7],
    3
   / \
  9  20
    /  \
   15   7
return its bottom-up level order traversal as:
[
  [15,7],
  [9,20],
  [3]
]
"""


# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def levelOrderBottom(self, root: TreeNode) -> List[List[int]]:
        result = [];
        if root is None: return result
        q=[]
        q.append(root)
        while len(q) > 0:
            nodes = []
            for i in range(len(q)):
                node = q.pop(0)
                nodes.append(node.val)
                if node.left != None: q.append(node.left)
                if node.right != None: q.append(node.right)
            
            result.insert(0,nodes)
        return result
